File: Sources/DanmakuOverlay/Filters/FilterEngine.swift
import Foundation

/// 弹幕屏蔽（需求文档 5.1.6 节：关键词 / 正则 / 颜色 / 长度 / 重复）
struct FilterRules {
    var keywords: [String] = []
    var regexPatterns: [String] = []
    var blockColored = false          // 屏蔽彩色弹幕（只留白色）
    var maxLength: Int = 0            // 0 表示不限制
    var mergeDuplicates = true        // 合并重复内容
    var showTop = true
    var showBottom = true
    var duplicateWindow: Double = 20  // 重复判定时间窗（秒）
}

enum FilterEngine {

    static func apply(_ list: [Danmaku], rules: FilterRules) -> [Danmaku] {
        let regexes = rules.regexPatterns.compactMap {
            try? NSRegularExpression(pattern: $0)
        }
        let keywords = rules.keywords.filter { !$0.isEmpty }

        var lastSeen: [String: Double] = [:]   // 文本 -> 上次出现时间
        var result: [Danmaku] = []
        result.reserveCapacity(list.count)

        for d in list {
            if d.mode == .other { continue }
            if !rules.showTop && d.mode == .top { continue }
            if !rules.showBottom && d.mode == .bottom { continue }
            if rules.blockColored && d.color != 0xFFFFFF { continue }
            if rules.maxLength > 0 && d.text.count > rules.maxLength { continue }
            if keywords.contains(where: { d.text.localizedCaseInsensitiveContains($0) }) { continue }

            if !regexes.isEmpty {
                let range = NSRange(d.text.startIndex..., in: d.text)
                if regexes.contains(where: { $0.firstMatch(in: d.text, range: range) != nil }) { continue }
            }

            if rules.mergeDuplicates {
                if let last = lastSeen[d.text], d.time - last < rules.duplicateWindow { continue }
                lastSeen[d.text] = d.time
            }

            result.append(d)
        }
        return result
    }

    /// 高密度降采样：每秒最多保留 maxPerSecond 条（按权重优先，此处按顺序截断）
    static func downsample(_ list: [Danmaku], maxPerSecond: Int) -> [Danmaku] {
        guard maxPerSecond > 0 else { return list }
        var counts: [Int: Int] = [:]
        return list.filter { d in
            let bucket = Int(d.time)
            let c = counts[bucket, default: 0]
            if c >= maxPerSecond { return false }
            counts[bucket] = c + 1
            return true
        }
    }
}

File: Windows/src/DanLaiDanQu.Core/FilterEngine.cs
using System.Text.RegularExpressions;

namespace DanLaiDanQu.Core;

public static class FilterEngine
{
    public static IReadOnlyList<Danmaku> Apply(IEnumerable<Danmaku> source, FilterRules rules)
    {
        var regexes = rules.RegexPatterns
            .Select(pattern => TryRegex(pattern))
            .Where(regex => regex is not null)
            .Cast<Regex>()
            .ToArray();
        var lastSeen = new Dictionary<string, double>(StringComparer.Ordinal);
        var result = new List<Danmaku>();

        foreach (var item in source)
        {
            if (item.Mode == DanmakuMode.Other ||
                (!rules.ShowTop && item.Mode == DanmakuMode.Top) ||
                (!rules.ShowBottom && item.Mode == DanmakuMode.Bottom) ||
                (rules.BlockColored && item.Color != 0xFFFFFF) ||
                (rules.MaxLength > 0 && item.Text.Length > rules.MaxLength) ||
                rules.Keywords.Any(keyword => item.Text.Contains(keyword, StringComparison.OrdinalIgnoreCase)) ||
                regexes.Any(regex => IsMatchSafe(regex, item.Text)))
            {
                continue;
            }

            if (rules.MergeDuplicates)
            {
                if (lastSeen.TryGetValue(item.Text, out var last) && item.Time - last < rules.DuplicateWindow)
                {
                    continue;
                }

                lastSeen[item.Text] = item.Time;
            }

            result.Add(item);
        }

        return result;
    }

    public static IReadOnlyList<Danmaku> Downsample(IEnumerable<Danmaku> source, int maxPerSecond)
    {
        if (maxPerSecond <= 0)
        {
            return source.ToArray();
        }

        var buckets = new Dictionary<int, int>();
        var result = new List<Danmaku>();
        foreach (var item in source)
        {
            var second = (int)item.Time;
            buckets.TryGetValue(second, out var count);
            if (count >= maxPerSecond)
            {
                continue;
            }

            buckets[second] = count + 1;
            result.Add(item);
        }

        return result;
    }

    public static (List<string> Keywords, List<string> Regexes) ParseFilterText(string text)
    {
        var keywords = new List<string>();
        var regexes = new List<string>();
        var token = new List<char>();
        var inRegex = false;
        var escaped = false;

        void AddToken()
        {
            var value = new string(token.ToArray()).Trim();
            token.Clear();
            if (value.Length == 0)
            {
                return;
            }

            if (value.Length > 2 && value.StartsWith('/') && value.EndsWith('/'))
            {
                regexes.Add(value[1..^1]);
            }
            else
            {
                keywords.Add(value);
            }
        }

        foreach (var character in text)
        {
            if ((character == ',' || character == '，') && !inRegex)
            {
                AddToken();
                escaped = false;
                continue;
            }

            if (character == '/' && !escaped)
            {
                inRegex = !inRegex;
            }

            token.Add(character);
            escaped = character == '\\' && !escaped;
            if (character != '\\')
            {
                escaped = false;
            }
        }

        AddToken();
        return (keywords, regexes);
    }

    public static IReadOnlyList<string> InvalidRegexPatterns(IEnumerable<string> patterns) => patterns
        .Where(pattern => TryRegex(pattern) is null)
        .ToArray();

    private static bool IsMatchSafe(Regex regex, string text)
    {
        try
        {
            return regex.IsMatch(text);
        }
        catch (RegexMatchTimeoutException)
        {
            // 用户规则不应使 UI 线程崩溃；超时内容按命中处理并屏蔽。
            return true;
        }
    }

    private static Regex? TryRegex(string pattern)
    {
        try
        {
            return new Regex(pattern, RegexOptions.IgnoreCase | RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100));
        }
        catch (ArgumentException)
        {
            return null;
        }
    }
}
